File: cliente.py
import socket
import msgpack


NOME_CLIENTE = input("nome do cliente:")  # Altere para "cliente2" no outro script


def publicar_mensagem():
    postagem = input("Digite sua postagem: ")

    payload = {
        "tipo": "postar",
        "origem": NOME_CLIENTE,
        "destino": "",
        "mensagem": postagem
    }

    soc.send(msgpack.packb(payload, use_bin_type=True))# comprimir o pacote de dado para ser enviada pela camada de rede
    resposta = msgpack.unpackb(soc.recv(1024), raw=False)   # o servidor nos retornara um resposta se ele receber algo, fiz essa interação para ficar algo legal e resposivo tipo o simbolo da nike quando a mensagem  no zap é enviada
    print("Status:", resposta.get("status")) # retornar o status

def vizualizar_postagens():
    origem = input("Digite com quem deseja vizualizar as postagens (ex: cliente1): ")
    payload = {
        "tipo": "vizualizar",
        "origem": origem
    }

    try:
        soc.send(msgpack.packb(payload, use_bin_type=True))
        resposta = soc.recv(4096)

        if not resposta:
            print("Nenhuma resposta do servidor.")
            return

        dados = msgpack.unpackb(resposta, raw=False)
        mensagens = dados.get("mensagens", [])

        if not mensagens:
            print("Nenhuma mensagem recebida.")
            return

        print("\nMensagens recebidas:\n")
        for i, msg in enumerate(mensagens, 1):
            origem = msg.get("origem", "desconhecido")
            texto = msg.get("mensagem", "")
            print(f"{i}. {origem} → {texto}")

    except Exception as e:
        print(f"Erro ao visualizar mensagens: {e}")


def enviar_mensagem_privada():
    destino = input("Digite o destinatário (ex: cliente2): ") # versatilidade para conseguir escolher o cliente a ser enviado por hora só foi feito entre o cliente 1 e 2
    mensagem = input("Digite sua mensagem: ")


    # chamei de payload por conta de um conceito de redes de computadores, não me aprofundarei no assunto
    payload = {
        "tipo": "enviar",
        "origem": NOME_CLIENTE,
        "destino": destino,
        "mensagem": mensagem
    }

    soc.send(msgpack.packb(payload, use_bin_type=True))# comprimir o pacote de dado para ser enviada pela camada de rede
    resposta = msgpack.unpackb(soc.recv(1024), raw=False)   # o servidor nos retornara um resposta se ele receber algo, fiz essa interação para ficar algo legal e resposivo tipo o simbolo da nike quando a mensagem  no zap é enviada
    print("Status:", resposta.get("status")) # retornar o status

def visualizar_mensagens():
    origem = input("Digite com quem deseja vizualizar a conversa (ex: cliente2): ")
    payload = {
        "tipo": "receber",
        "origem": "cliente2",
        "destino": NOME_CLIENTE
    }

    try:
        soc.send(msgpack.packb(payload, use_bin_type=True))
        resposta = soc.recv(4096)

        if not resposta:
            print("Nenhuma resposta do servidor.")
            return

        dados = msgpack.unpackb(resposta, raw=False)
        mensagens = dados.get("mensagens", [])

        if not mensagens:
            print("Nenhuma mensagem recebida.")
            return

        print("\nMensagens recebidas:\n")
        for i, msg in enumerate(mensagens, 1):
            origem = msg.get("origem", "desconhecido")
            texto = msg.get("mensagem", "")
            print(f"{i}. {origem} → {texto}")

    except Exception as e:
        print(f"Erro ao visualizar mensagens: {e}")


    except Exception as e:
        
        # fins de depuração
        print(f"Erro ao visualizar mensagens: {e}")

def menu():
    #fiz um menu interativo (ou quase isso) para usarmos em nosso projeto
    while True:
        print("\nMenu:")
        print("1 - Publicar mensagem")
        print("2 - Ver mensagens publicadas")
        print("3 - Enviar mensagem privada")
        print("4 - Visualizar mensagens privadas")
        print("0 - Sair")

        escolha = input("Escolha uma opção: ")

        if escolha == "1":
            publicar_mensagem()
        elif escolha == "2":
            vizualizar_postagens()
        elif escolha == "3":
            enviar_mensagem_privada()
        elif escolha == "4":
            visualizar_mensagens()
        elif escolha == "0":
            break
        else:
            print("Opção inválida.")

if __name__ == "__main__":
    HOST = '127.0.0.1'
    PORT = 5560

    soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    soc.connect((HOST, PORT))

    try:
        menu()
    finally:
        soc.close()

File: Postagens Demo/cliente.js
const net = require('net');
const msgpack = require('@msgpack/msgpack');
const readline = require('readline');

const HOST = '127.0.0.1';
const PORT = 65432;
const NOME_CLIENTE = "cliente1"; // Altere para "cliente2" em outro script

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const client = new net.Socket();

// Função auxiliar para ler entrada do usuário como Promise
function prompt(texto) {
    return new Promise(resolve => rl.question(texto, resolve));
}

async function publicarMensagem() {
    const postagem = await prompt("Digite sua mensagem: ");
        const payload = {
            tipo: "postar",
            origem: NOME_CLIENTE,
            destino: "",
            mensagem: postagem
        };

        client.write(msgpack.encode(payload), (err) => {
            if (err) {
                console.error("Erro ao enviar mensagem:", err);
                return;
            }

            client.once('data', (data) => {
                const resposta = msgpack.decode(data);
                console.log("Status:", resposta.status);
                menu();
            });
        });
    }


async function vizualizarPostagens() {
    const origem = await prompt("Digite o destinatário (ex: cliente2): ");

        const payload = {
            tipo: "vizualizar",
            origem: origem
        };

        client.write(msgpack.encode(payload), (err) => {
            if (err) {
                console.error("Erro ao enviar mensagem:", err);
                return;
            }

            client.once('data', (data) => {
                const resposta = msgpack.decode(data);
                const mensagens = resposta.mensagens || [];

            if (mensagens.length === 0) {
                console.log("Nenhuma mensagem recebida.");
                menu();
                return;
            }
            console.log("\nMensagens recebidas:\n");
            mensagens.forEach((msg, index) => {
                const msgOrigem = msg.origem || "desconhecido";
                const texto = msg.mensagem || "";
                console.log(`${index + 1}. ${msgOrigem} → ${texto}`);
            });

            menu();
        });
    });
}


// Enviar mensagem privada
async function enviar_mensagem_privada() {
    const destino = await prompt("Digite o destinatário (ex: cliente2): ");
    const mensagem = await prompt("Digite sua mensagem: ");

    const payload = {
        tipo: "enviar",
        origem: NOME_CLIENTE,
        destino,
        mensagem
    };

    client.write(msgpack.encode(payload));

    client.once('data', data => {
        const resposta = msgpack.decode(data);
        console.log("Status:", resposta.status);
        menu(); // volta pro menu
    });
}

// Visualizar mensagens recebidas
function visualizar_mensagens() {
    const payload = {
        tipo: "receber",
        destino: NOME_CLIENTE
    };

    client.write(msgpack.encode(payload));

    client.once('data', data => {
        try {
            const resposta = msgpack.decode(data);
            const mensagens = resposta.mensagens || [];

            if (mensagens.length === 0) {
                console.log("Nenhuma mensagem recebida.");
            } else {
                console.log("\nMensagens recebidas:\n");
                mensagens.forEach((msg, i) => {
                    const origem = msg.origem || "desconhecido";
                    const texto = msg.mensagem || "";
                    console.log(`${i + 1}. ${origem} → ${texto}`);
                });
            }
        } catch (err) {
            console.log("Erro ao visualizar mensagens:", err.message);
        }
        menu(); // volta pro menu
    });
}

// Menu interativo
async function menu() {
    console.log("\nMenu:");
    console.log("1 - Publicar mensagem (em construção)");
    console.log("2 - Enviar mensagem privada");
    console.log("3 - Visualizar mensagens privadas cliente 2");
    console.log("0 - Sair");

    const escolha = await prompt("Escolha uma opção: ");

    switch (escolha.trim()) {
        case '1':
            publicarMensagem();
            break;
        case '2':
            vizualizarPostagens();
            break;
        case "3":
            enviar_mensagem_privada();
            break;
        case "4":
            visualizar_mensagens();
            break;
        case "0":
            console.log("Encerrando...");
            client.end();
            rl.close();
            break;
        default:
            console.log("Opção inválida.");
            menu();
    }
}

// Conectar ao servidor e iniciar menu
client.connect(PORT, HOST, () => {
    console.log("Conectado ao servidor.");
    menu();
});
